File: ggts/providers/local_git.py
import os
import subprocess
from pathlib import Path
from typing import List, Optional
from datetime import datetime
from .base import BaseProvider
from ..models import CommitActivity
from ..parsers import TicketParser

class LocalGitProvider(BaseProvider):
    def __init__(self, base_dir: str):
        self.base_dir = base_dir

    def _find_git_repos(self, start_path: Path) -> List[Path]:
        repos = []
        for root, dirs, files in os.walk(start_path):
            if '.git' in dirs:
                repos.append(Path(root))
                # Halting recursion as requested: "Once a repository root is found, further recursion is halted"
                dirs.remove('.git') 
                dirs[:] = [] # Clear other dirs to prevent deep search
        return repos

    def fetch_commits(self, 
                      since: Optional[datetime] = None, 
                      until: Optional[datetime] = None,
                      author_patterns: Optional[List[str]] = None) -> List[CommitActivity]:
        commits: List[CommitActivity] = []
        repos = self._find_git_repos(Path(self.base_dir))
        
        for repo in repos:
            cmd = ['git', 'log', '--pretty=format:%H|%an|%ae|%aI|%s\n%b|||EOM|||']
            
            if since:
                cmd.append(f'--since={since.isoformat()}')
            if until:
                cmd.append(f'--until={until.isoformat()}')
                
            try:
                result = subprocess.run(cmd, cwd=repo, capture_output=True, text=True, check=True)
                raw_log = result.stdout
                
                # Split by custom delimiter
                entries = raw_log.split('|||EOM|||\n')
                for entry in entries:
                    entry = entry.strip()
                    if not entry:
                        continue
                        
                    parts = entry.split('|', 4)
                    if len(parts) < 5:
                        continue
                        
                    hash_str, author_name, author_email, date_str, message = parts
                    
                    # Filter by author if patterns provided
                    if author_patterns:
                        match = False
                        for pat in author_patterns:
                            pat_lower = pat.lower()
                            if pat_lower in author_name.lower() or pat_lower in author_email.lower():
                                match = True
                                break
                        if not match:
                            continue
                    
                    dt = datetime.fromisoformat(date_str)
                    tickets = TicketParser.extract_tickets(message)
                    
                    commits.append(CommitActivity(
                        commit_hash=hash_str,
                        author_name=author_name,
                        author_email=author_email,
                        timestamp=dt,
                        message=message.strip(),
                        project_name=repo.name,
                        provider='local_git',
                        tickets=tickets
                    ))
            except subprocess.CalledProcessError:
                print(f"Error reading git history in {repo}")
                
        return commits

File: src/providers/localGit.ts
import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';
import { BaseProvider } from './base';
import { CommitActivity } from '../models';
import { TicketParser } from '../parsers';

export class LocalGitProvider extends BaseProvider {
  baseDir: string;

  constructor(baseDir: string) {
    super();
    this.baseDir = baseDir;
  }

  private findGitRepos(dir: string, repos: string[] = []): string[] {
    try {
      const entries = fs.readdirSync(dir, { withFileTypes: true });
      for (const entry of entries) {
        if (entry.isDirectory()) {
          if (entry.name === '.git') {
            repos.push(dir);
            return repos; // Halt recursion as per spec
          } else if (entry.name !== 'node_modules' && entry.name !== 'python-legacy') {
            this.findGitRepos(path.join(dir, entry.name), repos);
          }
        }
      }
    } catch (e) {
      // Ignore permission or read errors
    }
    return repos;
  }

  async fetchCommits(since?: Date, until?: Date, authorPatterns?: string[]): Promise<CommitActivity[]> {
    const commits: CommitActivity[] = [];
    const repos = this.findGitRepos(this.baseDir).sort((a, b) => a.localeCompare(b));

    for (const repo of repos) {
      const projectName = path.basename(repo);
      let cmd = `git log --pretty=format:"%H|%an|%ae|%aI|%s%n%b|||EOM|||"`;
      
      if (since) cmd += ` --since="${since.toISOString()}"`;
      if (until) cmd += ` --until="${until.toISOString()}"`;

      try {
        const rawLog = execSync(cmd, { cwd: repo, encoding: 'utf-8', stdio: 'pipe' });
        const entries = rawLog.split('|||EOM|||\n');

        for (let entry of entries) {
          entry = entry.trim();
          if (!entry) continue;

          // Only split on the first 4 pipes, the rest belongs to the message
          const parts = entry.split('|');
          if (parts.length < 5) continue;

          const hash = parts[0];
          const authorName = parts[1];
          const authorEmail = parts[2];
          const dateStr = parts[3];
          const message = parts.slice(4).join('|').trim();

          if (authorPatterns && authorPatterns.length > 0) {
            let match = false;
            for (const pat of authorPatterns) {
              const pLower = pat.toLowerCase();
              if (authorName.toLowerCase().includes(pLower) || authorEmail.toLowerCase().includes(pLower)) {
                match = true;
                break;
              }
            }
            if (!match) continue;
          }

          const dt = new Date(dateStr);
          const tickets = TicketParser.extractTickets(message);

          commits.push({
            commitHash: hash,
            authorName,
            authorEmail,
            timestamp: dt,
            message,
            projectName,
            provider: 'local_git',
            durationMinutes: 0,
            tickets,
            isSessionStart: true
          });
        }
      } catch (e) {
        // Git repo has no commits yet or command failed
      }
    }

    return commits;
  }
}
